### ISSUE TITLE
Provide consult alternatives to `denote-insert-link` and `denote-org-extras-link-to-heading`. 

### ISSUE THREAD

**Issue by WeissP** [protesilaos/consult-denote#8]
It would be great if this package provides functions like `consult-denote-insert-link` and `consult-denote-link-to-heading` based on `consult-denote-find`.

**Comment by protesilaos** [comment]
> From: Weiss ***@***.***>
> Date: Sun, 20 Oct 2024 01:27:11 -0700
>
> It would be great if this package provides functions like
> `consult-denote-insert-link` and `consult-denote-link-to-heading`
> based on `consult-denote-find`.

Try to enable 'consult-denote-mode' and then call the standard Denote
commands. The idea is for those to be "powered up" behind the scenes
without redefining them.

That granted, the 'consult-denote-mode' is not related to
'consult-denote-find'. Do you want this in particular?

-- 
Protesilaos Stavrou
https://protesilaos.com

**Comment by WeissP** [comment]
Oh, sorry for the mix-up. What I meant was that I would like to use `consult-denote-find` to insert a denote link. The reason I'm interested in this is that I have thousands of notes, and using synchronized consult backends really slows things down, taking a few seconds every time I access them.

Before I started using `consult-denote`, I wrote my own consult-denote functions that show notes only under the current directory if it's within the `denote-directory`, or notes under a specific denote dir if it's currently outside of `denote-directory`. I also defined some narrow keys to switch between different denote sub directories. However, this requires dynamically updating the sources of `consult--multi` based on the current directory, which makes code ugly and hard maintain. Plus, I'm trying to avoid splitting notes into multiple directories, as it just adds extra confusion about where each note belongs.

On the other hand, keeping all my notes in one directory and using a synchronized consult backend has its own performance drawbacks. That’s why I want to experiment with `consult-denote-find`.

**Comment by protesilaos** [comment]
> From: Weiss ***@***.***>
> Date: Sun, 20 Oct 2024 06:13:45 -0700
>
> Oh, sorry for the mix-up. What I meant was that I would like to use
> `consult-denote-find` to insert a denote link. The reason I'm
> interested in this is that I have thousands of notes, and using
> synchronized consult backends really slows things down, taking a few
> seconds every time I access them.

Thank you for the clarification!

> Before I started using `consult-denote`, I wrote my own consult-denote
> functions that show notes only under the current directory if it's
> within the `denote-directory`, or notes under a specific denote dir if
> it's currently outside of `denote-directory`. I also defined some
> narrow keys to switch between different denote sub directories.
> However, this requires dynamically updating the sources of
> `consult--multi` based on the current directory, which makes code ugly
> and hard maintain. Plus, I'm trying to avoid splitting notes into
> multiple directories, as it just adds extra confusion about where each
> note belongs.

Do you think your code can be merged with what we have right now in
consult-denote? I am happy to cover more use-cases.

> On the other hand, keeping all my notes in one directory and using a
> synchronized consult backend has its own performance drawbacks. That’s
> why I want to experiment with `consult-denote-find`.

I am happy to have this done. What I do not remember though is whether
consult-find can be used just for selection without the side effect of
actually visiting the file. What we need it is to return an absolute
file path.

-- 
Protesilaos Stavrou
https://protesilaos.com

**Comment by WeissP** [comment]
> Do you think your code can be merged with what we have right now in consult-denote? I am happy to cover more use-cases.
 
Oh.. I really don't think my code can be merged into this package, I don't even want to maintain these functions anymore myself.

> What we need it is to return an absolute file path.

It seems like we have to use `consult--find` to get full path of found notes.

**Comment by protesilaos** [comment]
> From: Weiss ***@***.***>
> Date: Mon, 21 Oct 2024 01:09:32 -0700
>
>> Do you think your code can be merged with what we have right now in
>> consult-denote? I am happy to cover more use-cases.
>  
> Oh.. I really don't think my code can be merged into this package, I
> don't even want to maintain these functions anymore myself.

Fine, we will add new ones.

>> What we need it is to return an absolute file path.
>
> It seems like we have to use `consult--find` to get full path of found
> notes.

Does this work for you?

    (let ((dir (denote-directory)))
      (pcase-let* ((`(,prompt ,paths ,dir) (consult--directory-prompt "Find" dir))
                   (default-directory dir)
                   (builder (consult--find-make-builder paths)))
        (consult--find prompt builder nil)))

Or do you want the directory to always be the one you are in? If yes,
then we need a function that returns (i) the subdirectory of the
'denote-directory' or (ii) the 'denote-directory'.

-- 
Protesilaos Stavrou
https://protesilaos.com

**Comment by WeissP** [comment]
Yes that’s exactly what I’m looking for. Since this backend is asynchronous, I’m okay with searching through all the notes at once.

By the way, are you thinking about including the code you shared with me in this package? (You can go ahead and close this issue no matter the answer.)

**Comment by WeissP** [comment]
> Here is what I have right now

Looks nice!

> Which commands do you want to have here?

I mostly use commands `denote-insert-link` and `denote-org-extras-link-to-heading`.


A little bit off topic, do you think it is a good idea to make find program configurable? I personally prefer using `fd` instead of `find`.
